pennylane_lightning/core/src/simulators/lightning_tensor/tncuda/observables/ObservablesTNCuda.hpp

```diff
@@ -0,0 +1,543 @@
+        return std::shared_ptr<TensorProdObsTNCuda<StateTensorT>>{
+            new TensorProdObsTNCuda(std::move(obs))};
+    }
+
+    /**
+     * @brief Get the number of operations in observable.
+     *
+     * @return std::size_t
+     */
+    [[nodiscard]] auto getSize() const -> std::size_t { return obs_.size(); }
+
+    /**
+     * @brief Get the wires for each observable operation.
+     *
+     * @return const std::vector<std::size_t>&
+     */
+    [[nodiscard]] auto getWires() const -> std::vector<std::size_t> override {
+        return all_wires_;
+    }
+
+    /**
+     * @brief Get the name of the observable
+     */
+    [[nodiscard]] auto getObsName() const -> std::string override {
+        using Pennylane::Util::operator<<;
+        std::ostringstream obs_stream;
+        const auto obs_size = obs_.size();
+        for (size_t idx = 0; idx < obs_size; idx++) {
+            obs_stream << obs_[idx]->getObsName();
+            if (idx != obs_size - 1) {
+                obs_stream << " @ ";
+            }
+        }
+        return obs_stream.str();
+    }
+};
+
+/**
+ * @brief Hamiltonian representation as a sum of observables.
+ *
+ * @tparam StateTensorT State tensor class.
+ */
+template <class StateTensorT>
+class HamiltonianTNCuda : public ObservableTNCuda<StateTensorT> {
+  public:
+    using BaseType = ObservableTNCuda<StateTensorT>;
+    using PrecisionT = typename StateTensorT::PrecisionT;
+
+  private:
+    std::vector<PrecisionT> coeffs_ham_;
+    std::vector<std::shared_ptr<ObservableTNCuda<StateTensorT>>> obs_;
+
+    [[nodiscard]] bool
+    isEqual(const ObservableTNCuda<StateTensorT> &other) const override {
+        const auto &other_cast =
+            static_cast<const HamiltonianTNCuda<StateTensorT> &>(other);
+
+        if (coeffs_ham_ != other_cast.coeffs_ham_) {
+            return false;
+        }
+
+        for (size_t i = 0; i < obs_.size(); i++) {
+            if (*obs_[i] != *other_cast.obs_[i]) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+  public:
+    /**
+     * @brief Create a Hamiltonian from coefficients and observables
+     *
+     * @param coeffs Arguments to construct coefficients
+     * @param obs Arguments to construct observables
+     */
+    template <typename T1, typename T2>
+    HamiltonianTNCuda(T1 &&coeffs, T2 &&obs)
+        : coeffs_ham_{std::forward<T1>(coeffs)}, obs_{std::forward<T2>(obs)} {
+        BaseType::coeffs_ = coeffs_ham_;
+        PL_ASSERT(BaseType::coeffs_.size() == obs_.size());
+
+        for (std::size_t term_idx = 0; term_idx < BaseType::coeffs_.size();
+             term_idx++) {
+            auto ob = obs_[term_idx];
+            // This is aligned with statevector backends
+            PL_ABORT_IF(ob->getObsName().find("Hamiltonian") !=
+                            std::string::npos,
+                        "A Hamiltonian observable cannot be created from "
+                        "another Hamiltonian.");
```

Comment:
What about nested tensor-product observables? Is this supported?

Reply:
It's not supported as state vector backends.